<a href="https://colab.research.google.com/github/eunseochu/OB_5th/blob/main/%ED%8C%8C%EC%9D%B4%EC%8D%AC_%EB%A8%B8%EC%8B%A0%EB%9F%AC%EB%8B%9D_%EC%99%84%EB%B2%BD_%EA%B0%80%EC%9D%B4%EB%93%9C_%ED%85%8D%EC%8A%A4%ED%8A%B8%EB%B6%84%EC%84%9D(1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **8. 텍스트 분석**

### **1) NLP vs 텍스트 분석**
- NLP: 머신이 인간의 언어를 이해하고 해석하는 데 중점
- 텍스트 분석: 비정형 텍스트에서 의미 있는 정보를 추출하는 것에 중점

#### **NLP**
언어를 해석하기 위한 기계 번역, 자동으로 질문을 해석하고 답을 해주는 질의응답 시스템
- 텍스트 분석을 향상하게 하는 기반 기술
  - NLP 기술이 발전함에 따라 텍스트 분석도 더욱 정교하게 발전
- NLP와 텍스트 분석의 발전 근간에는 머신러닝이 존재
  - 이전에는 텍스트를 구성하는 언어적인 룰이나 업무의 룰에 따라 텍스트를 분석하는 룰 기반 시스템
  - 머신러닝의 텍스트 데이터를 기반으로 모델을 학습하고 예측하는 기반으로 변경되며 많은 기술적 발전 가능

#### **텍스트 분석**
모델을 수립하고 정보를 추출해 비즈니스 인텔리전스(BI)나 예측 분석등의 분석 작업을 주로 수행
- **텍스트 분류**: 문서가 특정 분류 또는 카테고리에 속하는 것을 예측하는 기법
  - ex. 특정 신문 기사 내용이 연애/정치/사회/문화 중 어떤 카테고리에 속하는지 자동으로 분류, 스팸 메일 검출
  - 지도 학습을 적용
- **감정 분석**: 텍스트에서 나타나는 감정/판단/믿음/의견/기분 등의 주관적인 요소를 분석하는 기법
  - ex. 소셜 미디어 감정 분석, 영화나 제품에 대한 긍정 또는 리뷰, 여론조사 의견 분석 등의 다양한 영역
  - 텍스트 분석에서 가장 활발하게 사용되는 분야이며, 지도학습 뿐만 아니라 비지도학습을 이용해 적용 가능
- **텍스트 요약**: 텍스트 내에서 중요한 주제나 중심 사상을 추출하는 기법
  - ex. 토픽 모델링
- **텍스트 군집화와 유사도 측정**: 비슷한 유형의 문서에 대해 군집화를 수행
  - 텍스트 분류를 비지도 학습으로 수행하는 방법의 일환
  - 유사도 측정 역시 문서들간의 유사도를 측정해 비슷한 문서끼리 모을 수 있는 방법

### **2) 텍스트 분석 이해**

#### **텍스트 분석이란?**
비정형 데이터인 텍스트를 분석하는 것
- ML 모델은 주어진 정형 데이터 기반에서 모델을 수립하고 예측을 수행하며, 숫자형의 피처 기반 데이터만 입력받을 수 있음
- 텍스트를 머신러닝에 적용하기 위해서는 비정형 텍스트 데이터를 어떻게 피처 형태로 추출하고 <br/> 추출된 피처에 의미 있는 값을 부여하는가 하는 것이 매우 중요한 요소
  - **피처 벡터화/추출**: 텍스트를 word 기반의 다수의 피처로 추출하고 <br/> 이 피처에 단어 빈도수와 같은 숫자 값을 부여하여 단어의 조합인 벡터값으로 표현
  - BOW(Bag of Words)와 Word2Vec 방법 존재

#### **텍스트 분석 수행 프로세스**
- **텍스트 사전 준비작업(텍스트 전처리)**
  - 텍스트를 피처로 만들기 전에 미리 클렌징
  - 대/소문자 변경, 특수 문자 삭제 등의 클렌징 작업, 대/소문자 변경, 특수문자 삭제 등의 클렌징 작업, <br/> 단어 등의 토큰화 작업, 의미 없는 단어 제거 작업, 어근 추출 등의 텍스트 정규화 작업 수행
- **피처 벡터화/추출**
  - 사전 준비 작업으로 가공된 텍스트에서 피처를 추출하고 여기에 벡터 값을 할당
  - BOW와 Word2Vec 방법 사용
  - BOW는 대표적으로 Count 기반과 TF-IDF 기반 벡터화 존재
- **ML 모델 수립 및 학습/예측/평가**
  - 피처 벡터화된 데이터 세트에 ML 모델을 적용해 학습/예측 및 평가를 수행

#### **파이썬 기반의 NLP, 텍스트 분석 패키지**
NLP와 텍스트 분석을 위해 쉽고 편하게 텍스트 사전 정제 작업, 피처 벡터화/추출, ML 모델을 지원하는 라이브러리 존재
- **NLTK**
방대한 데이터 세트와 서브 모듈을 가지고 있으며 NLP의 거의 모든 영역 커버
  - 많은 NLP 패키지가 NLTK의 영향을 받아 작성
  - 수행 성능과 정확도, 신기술, 엔터프라이즈한 기능 지원 등의 측면에서 부족한 부분 존재
  - 실제 대량의 데이터 기반에서는 제대로 활용되지 못하고 있음
- **Gensim**
  - 토픽 모델링 분야에서 가장 두각을 나타내는 패키지
  - 오래전부터 토픽 모델링을 쉽게 구현할 수 있는 기능을 제공해왔으며, Word2Vec 구현 등의 다양한 신기능 제공
  - SpaCy와 함께 실제 업무에서 자주 활용되는 NLP 패키지
- **Spacy**
  - 뛰어난 수행 성능으로 최근 가장 주목을 받는 NLP 패키지
  - 많은 NLP 어플리케이션에서 SpaCy를 사용하는 사례 증가

#### **사이킷런 라이브러리** 
사이킷런은 머신러닝 위주의 라이브러리여서 NLP를 위한 다양한 라이브러리는 가지고 있지 않음
- 텍스트를 일정 수준으로 가공하고 머신러닝 알고리즘에 텍스트 데이터를 피처로 처리하기 위한 편리한 기능 제공
- 사이킷런만으로도 충분한 텍스트 분석 기능을 수행 가능
- 더 다양한 텍스트 분석이 적용되어야하는 경우 NLP 전용 패키지와 결합해 애플리케이션을 작성

### **3) 텍스트 사전 준비 작업(텍스트 전처리) - 텍스트 정규화**
텍스트 자체를 바로 피처로 만들 수 없기 때문에, 사전에 텍스트를 가공하는 준비 작업이 필요
- 텍스트 정규화는 텍스트를 머신러닝 알고리즘이나 NLP 어플리케이션에 입력 데이터로 사용하기 위해 <br/> 클렌징, 정제, 토큰화, 어근화 등의 다양한 텍스트 데이터의 사전 작업을 수행하는 것을 의미
- 텍스트 분석은 텍스트 정규화 작업이 매우 중요
  - 클렌징
  - 토큰화
  - 필터링/스톱 워드 제거/철자 수정
  - Stemming
  - Lemmatizaion

#### **클렌징**
텍스트에서 분석에 오히려 방해가 되는 불필요한 문자, 기호 등을 사전에 제거하는 작업
- HTML, XML 태그나 특정 기호 등을 사전에 제거

#### **텍스트 토큰화**
문서에서 문장을 분리하는 문장 토큰화와 문장에서 단어를 토큰으로 분리하는 단어 토큰화로 구분
- NLTK는 이를 위해 다양한 API 제공

#### **문장 토큰화**
- 문장의 마침표(,), 개행문자(\n) 등 문장의 마지막을 뜻하는 기호에 따라 분리하는 것이 일반적
- 정규 표현식에 따른 문장 토큰화도 가능

**문장 토큰화 예제** <br/>
NLTK에서 일반적으로 많이 쓰이는 sent_tokenize를 이용해 토큰화를 수행
- 3개의 문장으로 이루어진 텍스트 문서를 문장으로 각각 분리하는 예제
- 단어 사전과 같이 참조가 필요한 데이터 세트의 경우 인터넷으로 다운로드
  - 수행하려는 컴퓨터에 인터넷 연결이 돼 있는지 확인 후 다운로드 수행
  - nltk.download('punkt')는 마침표, 개행 문자 등의 데이터 세트를 다운로드

In [2]:
from nltk import sent_tokenize
import nltk
nltk.download('punkt')

text_sample='The Matrix is everywhere its all around us, here even in ths room. You can see it out your window or on your television. You feel it when you go to work, or go to church or pay your taxes.'
sentences=sent_tokenize(text=text_sample)
print(type(sentences), len(sentences)) #list
print(sentences)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


<class 'list'> 3
['The Matrix is everywhere its all around us, here even in ths room.', 'You can see it out your window or on your television.', 'You feel it when you go to work, or go to church or pay your taxes.']


- sent_tokneize()가 반환하는 것은 각각의 문장으로 구성된 list 객체
- 반환된 list 객체가 3개의 문장으로 된 문자열을 가지고 있음을 알 수 있음

##### **단어 토큰화**
문장을 분리하는 구분자를 이용해 단어로 토큰화하는 것
- 공백, 콤마(,), 마침표(.), 개행문자 등으로 단어를 분리
- 정규 표현식을 이용해 다양한 유형으로 토큰화 수행 가능
  - Bag of Word와 같이 단어의 순서가 중요하지 않은 경우 단어 토큰화만 사용해도 충분
  - 문장 토큰화는 각 문장이 가지는 시맨틱적인 의미가 중요한 요소로 사용될 때 사용

**단어 토큰화 예제 (1)** <br/>
NTLK에서 기본으로 제공하는 word_tokenize()를 이용해 단어로 토큰화

In [3]:
from nltk import word_tokenize

sentence = "The Matrix is everywhere its all around us, here even in this room."
words = word_tokenize(sentence)
print(type(words), len(words))
print(words)

<class 'list'> 15
['The', 'Matrix', 'is', 'everywhere', 'its', 'all', 'around', 'us', ',', 'here', 'even', 'in', 'this', 'room', '.']


**단어 토큰화 예제(2)** <br/>
sent_tokenize와 word_tokenize를 조합해 문서에 대해서 모든 단어를 토큰화
- 이전 예제에서 선언된 3개의 문장으로 된 text_sample을 문장별로 단어 토큰화를 적용
- 문서를 먼저 문장으로 나누고, 개별 문장을 다시 단어로 토큰화하는 tokenize_text() 함수 생성

In [4]:
from nltk import word_tokenize, sent_tokenize

#여러개의 문장으로 된 입력 데이터를 문장별로 단어 토큰화 만드는 함수 생성
def tokenize_text(text):
    
    # 문장별로 분리 토큰
    sentences = sent_tokenize(text)
    # 분리된 문장별 단어 토큰화
    word_tokens = [word_tokenize(sentence) for sentence in sentences]
    return word_tokens

#여러 문장들에 대해 문장별 단어 토큰화 수행. 
word_tokens = tokenize_text(text_sample)
print(type(word_tokens),len(word_tokens))
print(word_tokens)

<class 'list'> 3
[['The', 'Matrix', 'is', 'everywhere', 'its', 'all', 'around', 'us', ',', 'here', 'even', 'in', 'ths', 'room', '.'], ['You', 'can', 'see', 'it', 'out', 'your', 'window', 'or', 'on', 'your', 'television', '.'], ['You', 'feel', 'it', 'when', 'you', 'go', 'to', 'work', ',', 'or', 'go', 'to', 'church', 'or', 'pay', 'your', 'taxes', '.']]


- 3개 문장을 문장별로 먼저 토큰화했으므로 word_tokens 변수는 3개의 리스트 객체를 내포하는 리스트
- 내포된 개별 리스트 객체는 각각 문장별로 토큰화된 단어를 요소로 가짐

**n-gram** <Br/>
**연속된 n개의 단어를 하나의 토큰화 단위로 분리**해 내는 것
- 문장을 단어별로 하나씩 토큰화할 경우 문맥적인 의미를 무시할 밖에 없는 문제를 해결하고자 도입
- n개 단어 크기 윈도우를 만들어 문장의 처음부터 오른쪽으로 움직이면서 토큰화를 수행

In [5]:
from nltk import ngrams

sentence = "The Matrix is everywhere its all around us, here even in this room."
words = word_tokenize(sentence)

all_ngrams = ngrams(words, 2)
ngrams = [ngram for ngram in all_ngrams]
print(ngrams)

[('The', 'Matrix'), ('Matrix', 'is'), ('is', 'everywhere'), ('everywhere', 'its'), ('its', 'all'), ('all', 'around'), ('around', 'us'), ('us', ','), (',', 'here'), ('here', 'even'), ('even', 'in'), ('in', 'this'), ('this', 'room'), ('room', '.')]


#### **스톱 워드 제거**
스톱 워드(Stop word)는 분석에 큰 의미가 없는 단어를 지칭
- is, the, a, will 등 문장을 구성하는 필수 문법 요소지만 문맥적으로 큰 의미가 없는 단어
- 문법적인 특성으로 인해 특히 빈번하게 텍스트에 나타나므로 **사전에 제거하지 않으면 중요한 단어로 인지될 수 있음**
- 의미 없는 단어를 제거하는 것이 중요한 전처리 작업
  - 언어별로 스톱 워드가 목록화되어 있음
  - NTLK는 가장 다양한 언어의 스톱 워드를 제공

##### **NLTK의 스톱워드**

**NLTK의 stopwords 목록 내려받기**

In [6]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

**NLTK English의 stopwords**
- 몇 개의 stopwrods가 있는지 알아보고 그 중 20개만 확인

In [7]:
print('영어 stop words 갯수:',len(nltk.corpus.stopwords.words('english')))
print(nltk.corpus.stopwords.words('english')[:40])

영어 stop words 갯수: 179
['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this']


- 스톱 워드의 개수가 179개이며 그 중 20개만 살펴보면 위 결과와 같음

**stopwords 필터링 제거** <br/>
3개의 문장별로 단어를 토큰화해 생성된 word_tokens 리스트에 대해 stopwords를 필터링으로 제거

In [8]:
import nltk

stopwords = nltk.corpus.stopwords.words('english')
all_tokens = []
# 위 예제의 3개의 문장별로 얻은 word_tokens list 에 대해 stop word 제거 Loop
for sentence in word_tokens:
    filtered_words=[]
    # 개별 문장별로 tokenize된 sentence list에 대해 stop word 제거 Loop
    for word in sentence:
        #소문자로 모두 변환합니다. 
        word = word.lower()
        # tokenize 된 개별 word가 stop words 들의 단어에 포함되지 않으면 word_tokens에 추가
        if word not in stopwords:
            filtered_words.append(word)
    all_tokens.append(filtered_words)
    
print(all_tokens)

[['matrix', 'everywhere', 'around', 'us', ',', 'even', 'ths', 'room', '.'], ['see', 'window', 'television', '.'], ['feel', 'go', 'work', ',', 'go', 'church', 'pay', 'taxes', '.']]


- is, this와 같은 스톱 워드가 필터링을 통해 제거 됨

#### **Stemming과 Lemmatization**
Stemming과 lemmatization은 문법적 또는 의미적으로 변화하는 단어의 원형을 찾음
- 원형 단어를 찾는다는 목적은 유사
- Lemmatization이 Stemming보다 정교하며 의미론적인 기반에서 단어의 원형을 찾음
  - Stemming: 원형 단어로 변환 시 일반적인 방법을 적용하거나 <br/> 더 단순화된 방법을 적용해 원래 단어에서 일부 철자가 훼손된 어근 단어를 추출하는 경향
  - Lemmatization: 품사와 같은 문법적인 요소와 더 의미적인 부분을 감안해 정확한 철자로 된 어근 단어를 찾으므로, <br/> Stemming보다 변환에 더 오랜 시간을 필요로 함

##### **NLTK의 Stemming과 Lemmatization**
NLTK는 다양한 Stemmer와 Lemmatizer 제공
- Stemmer: Porter, Lancaster, Snowball Stemmer
- Lemmatizer: WordNetLemmatizer

**NLTK의 LancasterStemmer** <br/>
진행형, 3인칭 단수, 과거형에 따른 동사, 그리고 비교, 최상에 따른 형용사의 변화에 따라 더 단순하게 원형 단어를 찾아줌
- LancasterStemmer()와 같이 필요한 Stemmer 객체를 생성
- **객체의 stem('단어') 메서드를 호출**하면 원하는 '단어'의 stemming 가능

In [9]:
from nltk.stem import LancasterStemmer
stemmer = LancasterStemmer()

print(stemmer.stem('working'),stemmer.stem('works'),stemmer.stem('worked'))
print(stemmer.stem('amusing'),stemmer.stem('amuses'),stemmer.stem('amused'))
print(stemmer.stem('happier'),stemmer.stem('happiest'))
print(stemmer.stem('fancier'),stemmer.stem('fanciest'))

work work work
amus amus amus
happy happiest
fant fanciest


- work의 경우 진행형, 3인칭 단수, 과거형 모두 기본 단어에 ing, s, ed가 붙는 단순한 변화이므로 원형 단어 인식
- amuse의 경우, 각 변화가 amus에 ing, s, ed가 붙으므로 정확한 단어 amuse가 아닌 amus를 원형 단어로 인식
- 형용사인 happy, fancy의 경우도 비교형 최상급형으로 변형된 단어의 정확한 원형을 찾지 못함
- 원형 단어에서 철자가 다른 어근 단어로 인식하는 경우가 발생

**NLTK의 WordNetLemmatizer** <br/>
Lemmatization은 보다 정확한 원형 단어 추출을 위해 **단어의 '품사'를 입력**해줘야 함
- lemmatize()의 파라미터로 동사의 경우 'v', 형용사의 경우 'a'를 입력

In [11]:
from nltk.stem import WordNetLemmatizer
import nltk
nltk.download('wordnet')

lemma = WordNetLemmatizer()
print(lemma.lemmatize('amusing', 'v'), lemma.lemmatize('amuses', 'v'), lemma.lemmatize('amused','v')) #동사 ->v
print(lemma.lemmatize('happier', 'a'), lemma.lemmatize('happiest', 'a')) #형용사 -> a
print(lemma.lemmatize('fancier', 'a'), lemma.lemmatize('happest', 'a'))

[nltk_data] Downloading package wordnet to /root/nltk_data...


amuse amuse amuse
happy happy
fancy happest


- 앞의 Stemmer보다 정확하게 원형 단어 추출

### **4) Bag of Words - BOW**
문서가 가지는 모든 단어(Words)를 문맥이나 순서를 무시하고 일괄적으로 **단어에 대해 빈도 값을 부여해 피처 값을 추출**하는 모델

#### **Bag of word의 단어 수(Word Count) 기반 피처 추출**
- 문장1과 2에 있는 모든 단어에서 중복을 제거하고 각 단어를 칼럼 형태로 나열. 그리고 각 단어에 고유의 인덱스 부여
  - 'and':0, 'baseball':1, 'daugter':2, 'games':3, ..., 'watch':9, 'wife':10
- 개별 문장에서 해당 단어가 나타나는 횟수를 각 단어에 기재

##### **BOW 모델의 장점**
- 단순히 단어의 발생 횟수에 기반하고 있지만, 예상보다 문서의 특징을 잘 나타낼 수 있는 모델
- 전통적으로 여러 분야에서 활용도가 높고, 쉽고 빠른 구축이 가능

##### **BOW 모델의 단점**
BOW 기반의 NLP 연구는 여러 가지 제약에 부딪히고 있음

**문맥 의미 반영 부족** <br/> 
단어의 순서를 고려하지 않기 때문에 문장 내에서 단어의 문맥적인 의미가 무시됨
  - n_gram 기법을 사용해 보완할 수 있지만, 제한적인 부분에 그침
  - 언어의 많은 부분을 차지하는 문맥적인 해석을 처리하지 못하는 단점

**희소 행렬 문제** <br/>
BOW로 피처 벡터화를 수행하면 희소 행렬 형태의 데이터 세트가 만들어지기 쉬움
- 많은 문서에서 단어를 추출하면 매우 많은 단어가 칼럼으로 만들어짐
  - 문서마다 서로 다른 단어가 구성되기에 단어가 문서마다 나타나지 않는 경우가 더 많음
  - 매우 많은 문서에서 단어의 총 개수는 수만 ~ 수십만개지만, <br/> 하나의 문서에 있는 단어는 이 중 극히 일부이므로 대부분의 데이터가 0으로 채워지게 됨
- **희소 행렬(Sparse Matrix)**: 대규모의 칼럼으로 구성된 행렬에서 **대부분의 값이 0으로 채워지는 행렬**
  - 이와 반대로 대부분의 값이 0이 아닌 의미 있는 값으로 채워져 있는 행렬은 밀집 행렬(Dense Matrix)
  - 일반적으로 **ML 알고리즘의 수행 시간과 예측 성능을 떨어트리기 때문에** 희소 행렬을 위한 특별한 기법 마련

#### **BOW 피처 벡터화**
머신러닝 알고리즘은 일반적으로 숫자형 피처를 데이터로 입력받아 동작하므로 텍스트와 같은 데이터는 바로 입력할 수 없음
- 텍스트를 **특정 의미를 가지는 숫자형 값인 벡터 값으로 변환하는 피처 벡터화**를 거쳐야 함
  - 각 문서의 텍스트를 단어로 추출해 피처로 할당하고, 각 단어의 발생 빈도와 같은 값을 피처에 값으로 부여해 <br/> 각 문서를 이 단어 피처의 발생 빈도 값으로 구성된 벡터로 만드는 기법
  - 기존 텍스트 데이터를 또 다른 형태의 피처의 조합으로 변경하기 때문에 넓은 의미의 피처 추출에 포함
- BOW 모델의 피처 벡터화는 모든 문서에서 모든 단어를 칼럼 형태로 나열하고 <br/> **각 문서에서 해당 단어의 횟수나 정규화된 빈도를 값으로 부여하는 데이터 세트 모델로 변경**하는 것
  - M개의 텍스트 문서가 있고, 이 문서에서 모든 단어를 추출해 나열했을 때 N개의 단어가 있다고 가정
  - 문서의 피처 벡터화를 수행하면 M개의 문서는 각각 N개의 값이 할당된 피처의 벡터 세트
  - M X N개의 단어 피처로 이뤄진 행렬을 구성

##### **BOW의 피처 벡터화 방식**
일반적으로 BOW의 피처 벡터화는 두 가지 방식 존재
- 카운트 기반의 벡터화
- TF_IDF(Term Frequency - Inverse Document Frequency) 기반의 벡터화

**카운트 벡터화** <br/>
단어 피처에 값을 부여할 때 각 문서에서 해당 단어가 나타나는 횟수, 즉 Count를 부여하는 경우
- 카운트 값이 높을수록 중요한 단어로 인식
- 카운트만 부여할 경우 언어의 특성상 문장에서 자주 사용될 수 밖에 없는 단어까지 높은 값을 부여하게 됨
- 이러한 문제를 보완하기 위해 TF-IDF 벡터화 사용

**TF-IDF 벡터화** <Br/>
개별 문서에서 자주 나타나는 단어에 높은 가중치를 주되, <br/> 모든 문서에서 전반적으로 자주 나타나는 단어에 대해서는 패널티를 주는 방식으로 값을 부여
- 단어가 단순히 등장하는 횟수에 따라 중요도를 평가받는다면 문서를 특정짓기가 어려워짐
  - 어떤 문서에서 특정 단어가 자주 나타나면 그 단어는 해당 문서를 특징짓는 중요 단어일 수 있음
  - 다른 문서에도 자주 나타나는 단어라면 해당 단어는 언어 특성상 범용적으로 자주 사용되는 단어일 가능성이 높음
- 모든 문서에서 반복적으로 자주 발생하는 단어에 대해서는 페널티를 부여해 단어에 대한 가중치의 균형을 맞춤
  - '많은', '빈번하게', '당연히' 등의 단어는 문서의 특징과 관련성이 적지만 보편적으로 많이 사용
  - 문서마다 텍스트가 길고 문서의 개수가 많은 경우 카운트 방식보다는 TF-IDF 방식이 더 좋은 예측 성능 보장

#### **사이킷런의 Count 및 TF-IDF 벡터화 구현: Countvectorizer, Tfidfvectorizer**

##### **사이킷런의 CountVectorizer 클래스**
사이킷런의 CountVectorizer 클래스는 카운트 기반의 벡터화를 구현한 클래스
- 피처 벡터화만을 수행하지는 않으며, 소문자 일괄 변환, 토큰화, 스톱 워드 필터링 등의 텍스트 전처리도 함께 수행
- 텍스트 전처리 및 피처 벡터화를 위한 입력 파라미터를 설정해 동작
- 다른 피처 변환 클래스와 마찬가지로 fit()과 transform()을 통해 피처 벡터화된 객체를 반환

**CountVectorizer의 입력 파라미터**
- **max_df**
  - 전체 문서에 걸쳐서 **너무 높은 빈도수를 가지는 단어 피처를 제외하기 위한 파라미터**
  - 너무 높은 빈도수를 가지는 단어는 스톱 워드와 비슷한 문법적인 특성으로 반복적인 단어일 가능성이 높기에 이를 제거
  - max_df = 100과 같이 정수 값을 가지면 전체 문서에 걸쳐 100개 이하로 나타나는 단어만 피처로 추출
  - max_df = 0.95와 같이 부동 소수점(0.0 ~ 1.0) 값을 가지면, <br/> 전체 문서에 걸쳐 빈도수 0~95% 까지의 단어만 피처로 추출하고 나머지 상위 5%는 피처로 추출하지 않음
- **min_df**
  - 전체 문서에 걸쳐서 **너무 낮은 빈도수를 가지는 단어 피처를 제외하기 위한 파라미터**
  - 수백 ~ 수천 개의 전체 문서에서 특정 단어가 min_df에 설정된 값보다 적은 빈도수를 가진다면, <br/> 이 단어는 크게 중요하지 않거나 가비지(garbage)성 단어일 확률이 높음
  - min_df = 2와 같이 정수 값을 가지면 전체 문서에 걸쳐 2번 이하로 나타나는 단어는 피처로 추출하지 않음
  - min_df = 0.02와 같이 부동 소수점(0.0 ~ 1.0) 값을 가지면, <br/> 전체 문서에 걸쳐 하위 2% 이하의 빈도수를 가지는 단어는 피처로 추출하지 않음
- **max_features**
  - **추출하는 피처의 개수를 제한**하며 정수로 값을 지정
  - max_features = 2000으로 지정할 경우 가장 높은 빈도를 가지는 단어 순으로 정렬해 2000개까지만 피처로 추출
- **stop_words**
  - 'english'로 지정하면 영어의 스톱 워드로 지정된 단어는 추출에서 제외
- **n_gram_range**
  - BOW 모델의 단어 순서를 어느 정도 보강하기 위핸 **n_gram 범위를 설정**
  - 튜플 형태로 (범위 최솟값, 범위 최댓값)을 지정
  - (1, 1)로 지정하면 토큰화된 단어를 1개씩 피처로 추출
  - (1, 2)로 지정하면 토큰화된 단어를 1개씩, 그리고 순서대로 2개씩 묶어서 피처로 추출
- **analyzer**
  - 피처 추출을 수행한 단위를 지정하며, 디폴트는 'word'
  - word가 아니라 character의 특정 범위를 피처로 만드는 특정한 경우 등을 적용할 때 사용
- **token_pattern**
  - 토큰화를 수행하는 정규 표현식 패턴을 지정하며, 디폴트는 '\b\w\w+\b'
  - 공백 또는 개행 문자 등으로 구분된 단어 분리자 사이의 2문자(문자 또는 숫자, 등 영숫자) 이상의 단어를 토큰으로 분리
  - analyzer = word로 설정했을 때만 변경 가능하나, 거의 변경하지 않음
- **tokenizer**
  - 토큰화를 별도의 커스텀 함수로 이용 시 적용
  - 일반적으로 CountTokenizer 클래스에서 어근 변환 시 이를 수행하는 별도의 함수를 파라미터로 적용

**CountVectorizer를 이용한 피처 벡터화 방법** <br/>
사이킷런의 CountVectorizer 클래스를 이용해 카운트 기반의 피처 여러 개의 문서로 구성된 텍스트의 피처 벡터화 방법
- **사전 데이터 가공**
  - 영어의 경우 모든 문자를 소문자로 변경하는 등의 전처리 작업을 수행
  - Default로 lowercase = True
- **토큰화**
  - n_gram_range를 반영해 각 단어를 토큰화
  - Default로 단어 기준 (analyzer = True)
- **텍스트 정규화**
  - stop_wrods = 'english'와 같이 주어진 경우 **stop words 필터링만 수행 가능**
  - Stemming과 Lmmatization 같은 어근 변환은 CountBectorizer에서 직접 지원하지 않음
  - tokenizer 파라미터에서 커스텀 어근 변환 함수를 적용하거나 외부패키지로 Text Normalization 수행
- **피처 벡터화**
  - max_df, min_df, max_features 등 파라미터를 이용해 토큰화된 단어로 피처를 추출하고, 단어 빈도수 벡터 값을 적용

##### **사이킷런의 TfidfVectorizer 클래스**
사이킷런의 TfidfVectorizer 클래스는 TF-IDF 벡터화를 구현한 클래스
- 파라미터와 변환 방법은 CountVectorizer와 동일

#### **BOW 벡터화를 위한 희소 행렬**
사이킷런의 CountVectorizer/TfdfVectorizer를 이용해 텍스트를 피처 단위로 벡터화해 변환하고 CSR 형태의 희소 행렬 반환
- 사용자 입장에서 피처 벡터화된 희소 행렬이 어떤 형태인지 중요하지 않을 수 있음
- 좀 더 난이도 있는 ML 모델을 수립하기 위해서는 이러한 희소 행렬이 어떤 형태로 돼 있는지 알아야 함

**희소 행렬** <br/>
대규모 행렬의 대부분의 값을 0이 차지하는 행렬
- BOW 형태를 가진 언어 모델의 피처 벡터화는 대부분 희소 행렬

**희소 행렬의 방식**
- COO 형식과 CSR 형식이 존재

#### **희소 행렬 - COO 형식 (Coordinate: 좌표)**
0이 아닌 데이터만 별도의 데이터 배열에 저장하고, 그 데이터가 가르키는 행과 열의 위치를 별도의 배열로 저장하는 방식

In [13]:
import numpy as np

dense = np.array( [ [ 3, 0, 1 ], 
                    [0, 2, 0 ] ] )

In [14]:
from scipy import sparse

# 0 이 아닌 데이터 추출
data = np.array([3,1,2])

# 행 위치와 열 위치를 각각 array로 생성 
row_pos = np.array([0,0,1])
col_pos = np.array([0,2,1])

# sparse 패키지의 coo_matrix를 이용하여 COO 형식으로 희소 행렬 생성
sparse_coo = sparse.coo_matrix((data, (row_pos,col_pos)))

In [15]:
print(type(sparse_coo))
print(sparse_coo)
dense01=sparse_coo.toarray()
print(type(dense01),"\n", dense01)

<class 'scipy.sparse._coo.coo_matrix'>
  (0, 0)	3
  (0, 2)	1
  (1, 1)	2
<class 'numpy.ndarray'> 
 [[3 0 1]
 [0 2 0]]


#### **희소 행렬 - CSR 형식 (Compressed Sparse Row)** <br/>
COO 형식이 행과 열의 위치를 나타내기 위해서 반복적인 위치 데이터를 사용해야 하는 문제점을 해결한 방식

In [16]:
from scipy import sparse

dense2 = np.array([[0,0,1,0,0,5],
             [1,4,0,3,2,5],
             [0,6,0,3,0,0],
             [2,0,0,0,0,0],
             [0,0,0,7,0,8],
             [1,0,0,0,0,0]])

# 0 이 아닌 데이터 추출
data2 = np.array([1, 5, 1, 4, 3, 2, 5, 6, 3, 2, 7, 8, 1])

# 행 위치와 열 위치를 각각 array로 생성 
row_pos = np.array([0, 0, 1, 1, 1, 1, 1, 2, 2, 3, 4, 4, 5])
col_pos = np.array([2, 5, 0, 1, 3, 4, 5, 1, 3, 0, 3, 5, 0])

# COO 형식으로 변환 
sparse_coo = sparse.coo_matrix((data2, (row_pos,col_pos)))

# 행 위치 배열의 고유한 값들의 시작 위치 인덱스를 배열로 생성
row_pos_ind = np.array([0, 2, 7, 9, 10, 12, 13])

# CSR 형식으로 변환 
sparse_csr = sparse.csr_matrix((data2, col_pos, row_pos_ind))

print('COO 변환된 데이터가 제대로 되었는지 다시 Dense로 출력 확인')
print(sparse_coo.toarray())
print('CSR 변환된 데이터가 제대로 되었는지 다시 Dense로 출력 확인')
print(sparse_csr.toarray())

COO 변환된 데이터가 제대로 되었는지 다시 Dense로 출력 확인
[[0 0 1 0 0 5]
 [1 4 0 3 2 5]
 [0 6 0 3 0 0]
 [2 0 0 0 0 0]
 [0 0 0 7 0 8]
 [1 0 0 0 0 0]]
CSR 변환된 데이터가 제대로 되었는지 다시 Dense로 출력 확인
[[0 0 1 0 0 5]
 [1 4 0 3 2 5]
 [0 6 0 3 0 0]
 [2 0 0 0 0 0]
 [0 0 0 7 0 8]
 [1 0 0 0 0 0]]


In [17]:
print(sparse_csr)

  (0, 2)	1
  (0, 5)	5
  (1, 0)	1
  (1, 1)	4
  (1, 3)	3
  (1, 4)	2
  (1, 5)	5
  (2, 1)	6
  (2, 3)	3
  (3, 0)	2
  (4, 3)	7
  (4, 5)	8
  (5, 0)	1


In [18]:
dense3 = np.array([[0,0,1,0,0,5],
             [1,4,0,3,2,5],
             [0,6,0,3,0,0],
             [2,0,0,0,0,0],
             [0,0,0,7,0,8],
             [1,0,0,0,0,0]])

coo = sparse.coo_matrix(dense3)
csr = sparse.csr_matrix(dense3)